core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,264 @@
+
+    val updatedAcls: Set[Acl] = getAcls(resource) ++ acls
+    val path: String = toResourcePath(resource)
+
+    if (ZkUtils.pathExists(zkClient, path))
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+    else
+      ZkUtils.createPersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, updatedAcls)
+    }
+
+    updateAclChangedFlag(resource)
+  }
+
+  override def removeAcls(acls: Set[Acl], resource: Resource): Boolean = {
+    if (!ZkUtils.pathExists(zkClient, toResourcePath(resource)))
+      return false
+
+    val existingAcls: Set[Acl] = getAcls(resource)
+    val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !acls.contains(acl))
+    if (existingAcls.equals(filteredAcls))
+      return false
+
+    val path: String = toResourcePath(resource)
+    ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, filteredAcls)
+    }
+
+    updateAclChangedFlag(resource)
+
+    true
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    if (ZkUtils.pathExists(zkClient, toResourcePath(resource))) {
+      ZkUtils.deletePath(zkClient, toResourcePath(resource))
+      inWriteLock(lock) {
+        aclCache.remove(resource)
+      }
+      updateAclChangedFlag(resource)
+      true
+    } else false
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      if (aclCache.contains(resource))
+        return aclCache.get(resource).get
+    }
+
+    val aclJson: Option[String] = ZkUtils.readDataMaybeNull(zkClient, toResourcePath(resource))._1
+    val acls: Set[Acl] = aclJson map ((x: String) => Acl.fromJson(x)) getOrElse Set.empty
+
+    inWriteLock(lock) {
+      aclCache.put(resource, acls)
+    }
+    acls
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Set[Acl] = {
+    val aclSets = aclCache.values
+    var acls = Set.empty[Acl]
+    for(aclSet: Set[Acl]  <- aclSets) {
+      acls ++= aclSet.filter(acl => acl.principal.equals(principal))
+    }
+    acls.toSet[Acl]
+  }
+
+  def loadCache: Set[Acl] = {
+    var acls = Set.empty[Acl]
+    val resourceTypes = ZkUtils.getChildren(zkClient, SimpleAclAuthorizer.AclZkPath)
+    for (rType <- resourceTypes) {
+      val resourceType = ResourceType.fromString(rType)
+      for (resources <- ZkUtils.getChildren(zkClient, resourceType.name)) {
+        for (resourceName <- resources)
+          acls ++= getAcls(Resource(resourceType, resourceName.toString))
+      }
+    }
+    acls
+  }
+
+
+  def toResourcePath(resource: Resource): String = {
+    SimpleAclAuthorizer.AclZkPath + "/" + resource.resourceType + "/" + resource.name
+  }
+
+  def updateAclChangedFlag(resource: Resource): Unit = {
+      ZkUtils.updatePersistentPath(zkClient, SimpleAclAuthorizer.AclChangedZkPath, resource.toString)
```

Comment:
Hmm, this may not work well if two ACL changes happen quickly. The second one will override the the resource of the previous one. If a broker hasn't finished processing the previous change, it will miss it after the override. I thought we wanted to use the same approach as in ConfigChangeListener where the updates are written as a sequential ZK node.